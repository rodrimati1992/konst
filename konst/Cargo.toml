[package]
name = "konst"
version = "0.3.5"
authors = ["rodrimati1992 <rodrimatt1985@gmail.com>"]
rust-version = "1.65.0"
edition = "2021"
license = "Zlib"
description = "Const equivalents of std functions, compile-time comparison, and parsing"
documentation = "https://docs.rs/konst/"
readme="../README.md"
keywords = ["no-std", "const_fn", "parsing"]
categories = ["no-std", "parsing"]
repository = "https://github.com/rodrimati1992/konst/"
include = [
    "Cargo.toml", 
    "src/**/*.rs", 
    "../README.md",
    "LICENSE-ZLIB.md", 
]


[dependencies.konst_proc_macros]
version = "=0.3.0"
path = "../konst_proc_macros"
optional = true

[dependencies.konst_kernel]
version = "=0.3.5"
path = "../konst_kernel"
features = ["rust_1_64", "__for_konst"]

[dev-dependencies.rand]
version = "0.8.4"
default-features = false
features = ["small_rng"]

# dev-dependencies can't be optional
[dependencies.trybuild]
version = "1.0"
optional = true

[dependencies.typewit]
version = "1.1"
features = ["rust_1_61"]

[dependencies.const_panic]
version = "0.2.6"
features = ["rust_1_64"]
default-features = false

[features]
default = ["cmp", "iter", "parsing_proc"]


# Enables extra checks for debug
debug = ["konst_kernel/debug"]
cmp = []
iter = ["konst_kernel/iter"]
parsing_proc = ["parsing", "konst_proc_macros"]
parsing = []
rust_latest_stable = []
mut_refs = ["rust_latest_stable", "konst_kernel/mut_refs"]
nightly_mut_refs = ["mut_refs", "konst_kernel/nightly_mut_refs"]

alloc = ["konst_kernel/alloc"]

# Enables ui tests, which are intended for the latest stable release.
# This is not enabled in CI, because it's way too sensitive to 
# changes in how rustc formats errors
__ui = ["trybuild", "rust_latest_stable"]
__cp_derive = ["const_panic/derive", "const_panic/non_basic"]

docsrs = ["konst_kernel/docsrs"]



[package.metadata.docs.rs]
features = ["docsrs", "alloc", "cmp", "parsing_proc", "rust_latest_stable", "nightly_mut_refs"]

