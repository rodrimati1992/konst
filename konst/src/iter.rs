//! Const equivalent of iterators with a specific `next` function signature.
//!
//! The docs for [`ConstIntoIter`] has more information on
//! const equivalents of IntoIterator and Iterator.
//!

mod iterator_adaptors;
pub mod iterator_dsl;

pub use iterator_adaptors::*;

/// Iterates over all elements of an [iterator](crate::iter::ConstIntoIter),
/// const equivalent of [`Iterator::for_each`]
///
/// # Syntax
///
/// ```text
/// for_each!{
///     $pattern:pat in $iterator:expr
///         $(,$iterator_method:ident ($($method_args:tt)*) )*
///         $(,)?
///     =>
///     $($code:tt)*
/// }
/// ```
///
/// This macro supports emulating iterator methods by expanding to equivalent code.
/// They are documented in the [`iterator_dsl`] module,
/// because they are also supported by other `konst::iter` macros.
///
/// # Examples
///
/// ### Custom iterator
///
/// ```rust
/// use konst::iter::{ConstIntoIter, IsIteratorKind};
///
/// struct Upto10(u8);
///
/// impl ConstIntoIter for Upto10 {
///     type Kind = IsIteratorKind;
///     type IntoIter = Self;
///     type Item = u8;
/// }
///
/// impl Upto10 {
///     const fn next(mut self) -> Option<(u8, Self)> {
///         if self.0 < 10 {
///             let ret = self.0;
///             self.0 += 1;
///             Some((ret, self))
///         } else {
///             None
///         }
///     }
/// }
///
/// const N: u32 = {
///     let mut n = 0u32;
///     konst::iter::for_each!{elem in Upto10(7) =>
///         n = n * 10 + elem as u32;
///     }
///     n
/// };
///
/// assert_eq!(N, 789);
///
/// ```
///
/// ### Summing pairs
///
/// ```rust
/// use konst::iter::for_each;
///     
/// const fn add_pairs<const N: usize>(l: [u32; N], r: [u32; N]) -> [u32; N] {
///     let mut out = [0u32; N];
///
///     for_each!{(i, val) in &l,zip(&r),map(|(l, r)| *l + *r),enumerate() =>
///         out[i] = val;
///     }
///
///     out
/// }
///
/// assert_eq!(add_pairs([], []), []);
/// assert_eq!(add_pairs([3], [5]), [8]);
/// assert_eq!(add_pairs([3, 5], [8, 13]), [11, 18]);
///
/// ```
///
/// [`iterator_dsl`]: crate::iter::iterator_dsl
pub use konst_kernel::for_each;

/// Wrapper for `ConstIntoIter` implementors,
/// that defines different methods depending on the
/// value of `K`.
#[doc(inline)]
pub use konst_kernel::into_iter::IntoIterWrapper;

/// Marker type for proving that `T: ConstIntoIter<Kind = K>`
#[doc(inline)]
pub use konst_kernel::into_iter::IsConstIntoIter;

/// Macro for converting [`ConstIntoIter`] implementors into const iterators.
///
/// # Behavior
///
/// For std types (`ConstIntoIter<Kind = IsStdKind>`),
/// this converts those types to their iterator.
/// [(example below)](#std-type-example)
///
/// For user-defined into-iterators (`ConstIntoIter<Kind = IsIntoIterKind>`),
/// this calls their `const_into_iter` inherent method to convert them to an iterator.
/// [(example below)](#into-iter-example)
///
/// For iterators (`ConstIntoIter<Kind = IsIteratorKind>`),
/// this returns the iterator untouched.
/// [(example below)](#iterator-example)
///
/// # Examples
///
/// <span id="std-type-example"></span>
/// ### Std type
///
/// This example demonstrates passing a `ConstIntoIter<Kind = IsStdKind>` in.
///
/// ```rust
/// use konst::{iter, slice};
///
/// let mut elem;
/// let mut iter: slice::Iter<'_, u8> = iter::into_iter!(&[3, 5, 8]);
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, &3);
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, &5);
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, &8);
///
/// assert!(iter.next().is_none());
///
/// ```
///
/// <span id="into-iter-example"></span>
/// ### IntoIterator type
///
/// This example demonstrates passing a `ConstIntoIter<Kind = IsIntoIterKind>` in.
///
/// ```rust
/// use konst::{iter, string};
///
/// let mut iter: Countdown = iter::into_iter!(Number(3));
/// let mut elem;
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, 2);
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, 1);
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, 0);
///
/// assert!(iter.next().is_none());
///
///
/// struct Number(u32);
///
/// impl iter::ConstIntoIter for Number {
///     type Kind = iter::IsIntoIterKind;
///     type Item = u32;
///     type IntoIter = Countdown;
/// }
///
/// impl Number {
///     const fn const_into_iter(self) -> Countdown {
///         Countdown(self.0)
///     }
/// }
///
/// struct Countdown(u32);
///
/// impl iter::ConstIntoIter for Countdown {
///     type Kind = iter::IsIteratorKind;
///     type Item = u32;
///     type IntoIter = Self;
/// }
///
/// impl Countdown {
///     const fn next(self) -> Option<(u32, Self)> {
///         let next = konst::try_opt!(self.0.checked_sub(1));
///         Some((next, Countdown(next)))
///     }
/// }
///
///
/// ```
///
/// <span id="iterator-example"></span>
/// ### Iterator type
///
/// This example demonstrates passing a `ConstIntoIter<Kind = IsIteratorKind>` in.
///
/// ```rust
/// use konst::{iter, string};
///
/// let iter: string::Split<'_, '_, char> = string::split("foo bar baz", ' ');
///
/// // `iter::into_iter` is an identity function when passed iterators
/// let mut iter: string::Split<'_, '_, char> = iter::into_iter!(iter);
/// let mut elem;
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, "foo");
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, "bar");
///
/// (elem, iter) = iter.next().unwrap();
/// assert_eq!(elem, "baz");
///
/// assert!(iter.next().is_none());
/// ```
///
///
#[doc(inline)]
pub use konst_kernel::into_iter_macro as into_iter;

/// Const analog of the [`IntoIterator`] trait.
///
/// # Implementor
///
/// Implementors are expected to be one of these:
///
/// - [`IsIntoIterKind` kind](#isintoiterkind)
/// - [`IsIteratorKind` kind](#isiteratorkind)
/// - Standard library types, of the [`IsStdKind`] kind
///
/// ### `IsIntoIterKind`
///
/// These are user-defined types convertible to const iterators.
///
/// These implement `ConstIntoIter<Kind = `[`IsIntoIterKind`]`>`
/// and are expected to define this inherent method for converting to
/// a const iterator:
///
/// ```rust
/// # struct II;
/// # struct SomeIterator;
/// # impl II {
/// const fn const_into_iter(self) -> SomeIterator {
/// #   loop{}
/// # }
/// # }
/// ```
///
/// [full example below](#non-iter-example)
///
/// ### `IsIteratorKind`
///
/// These are const iterator types.
///
/// These implement `ConstIntoIter<Kind = `[`IsIteratorKind`]`>`
/// and are expected to define this inherent method:
///
/// ```rust
/// # struct SomeIterator;
/// # type Item = u8;
/// # impl SomeIterator {
/// // Equivalent to `Iterator::next`
/// const fn next(self) -> Option<(Item, Self)> {
/// #   loop{}
/// # }
/// # }
/// ```
/// Where `Item` can be any type.
///
/// These are other methods that you can optionaly define,
/// which most iterators from the `konst` crate define:
/// ```rust
/// # struct SomeIterator;
/// # struct SomeIteratorRev;
/// # type Item = u8;
/// # impl SomeIterator {
/// // equivalent to `DoubleEndedÃŒterator::mext_back`
/// const fn next_back(self) -> Option<(Item, Self)> {
/// #   loop{}
///     // ... some code...
/// }
///
/// // Reverses the iterator, equivalent to `Iterator::rev`
/// const fn rev(self) -> SomeIteratorRev {
/// #   loop{}
///     // ... some code...
/// }
///
/// // Clones the iterator, equivalent to `Clone::clone`
/// const fn copy(&self) -> Self {
/// #   loop{}
///     // ... some code...
/// }
/// # }
/// ```
/// Where `SomeIteratorRev` should be a `ConstIntoIter<Kind = IsIteratorKind>`
/// which has the same inherent methods for iteration,
/// and returns the same `Item` type.
///
/// [full example below](#iter-example)
///
/// # Examples
///
/// <span id = "non-iter-example"></span>
/// ### Implementing for an into-iterator
///
/// ```rust
/// use konst::{iter, slice};
///
/// struct GetSlice<'a, T>{
///     slice: &'a [T],
///     up_to: usize,
/// }
///
/// impl<'a, T> iter::ConstIntoIter for GetSlice<'a, T> {
///     type Kind = iter::IsIntoIterKind;
///     type IntoIter = konst::slice::Iter<'a, T>;
///     type Item = &'a T;
/// }
///
/// impl<'a, T> GetSlice<'a, T> {
///     const fn const_into_iter(self) -> konst::slice::Iter<'a, T> {
///         slice::iter(slice::slice_up_to(self.slice, self.up_to))
///     }
/// }
///
/// const fn sum_powers(up_to: usize) -> u64 {
///     let gs = GetSlice{slice: &[1, 2, 4, 8, 16, 32, 64, 128], up_to};
///
///     iter::eval!(gs,fold(0, |l, &r| l + r))
/// }
///
/// assert_eq!(sum_powers(0), 0);
/// assert_eq!(sum_powers(1), 1);
/// assert_eq!(sum_powers(2), 3);
/// assert_eq!(sum_powers(3), 7);
/// assert_eq!(sum_powers(4), 15);
/// assert_eq!(sum_powers(5), 31);
///
/// ```
///
/// <span id = "iter-example"></span>
/// ### Implementing for an iterator
///
/// ```rust
/// use konst::iter::{self, ConstIntoIter};
///
/// struct Countdown(u8);
///
/// impl ConstIntoIter for Countdown {
///     type Kind = iter::IsIteratorKind;
///     type IntoIter = Self;
///     type Item = u8;
/// }
///
/// impl Countdown {
///     const fn next(mut self) -> Option<(u8, Self)> {
///         konst::option::map!(self.0.checked_sub(1), |ret| {
///             self.0 = ret;
///             (ret, self)
///         })
///     }
/// }
///
/// const fn sum(initial: u8) -> u16 {
///     iter::eval!(Countdown(initial),fold(0u16, |accum, elem| accum + elem as u16))
/// }
///
/// assert_eq!(sum(0), 0);
/// assert_eq!(sum(1), 0);
/// assert_eq!(sum(2), 1);
/// assert_eq!(sum(3), 3);
/// assert_eq!(sum(4), 6);
/// assert_eq!(sum(5), 10);
///
/// ```
///
/// ### Implementing for a double-ended iterator
///
/// ```rust
/// use konst::iter;
///
/// assert_eq!(HOURS, [1, 2, 3, 4, 5, 6, 12, 11, 10, 9, 8, 7]);
///
/// const HOURS: [u8; 12] = {
///     let mut arr = [0; 12];
///     let hours = Hours::new();
///
///     iter::for_each!{(i, hour) in 0..6,zip(hours.copy()) =>
///         arr[i] = hour;
///     }
///     iter::for_each!{(i, hour) in 6..12,zip(hours.rev()) =>
///         arr[i] = hour;
///     }
///
///     arr
/// };
///
///
/// struct Hours{
///     start: u8,
///     end: u8,
/// }
///
/// impl iter::ConstIntoIter for Hours {
///     type Kind = iter::IsIteratorKind;
///     type IntoIter = Self;
///     type Item = u8;
/// }
///
/// impl Hours {
///     const fn new() -> Self {
///         Self {start: 1, end: 13}
///     }
///
///     const fn next(mut self) -> Option<(u8, Self)> {
///         if self.start == self.end {
///             None
///         } else {
///             let ret = self.start;
///             self.start += 1;
///             Some((ret, self))
///         }
///     }
///
///     const fn next_back(mut self) -> Option<(u8, Self)> {
///         if self.start == self.end {
///             None
///         } else {
///             self.end -= 1;
///             Some((self.end, self))
///         }
///     }
///
///     const fn rev(self) -> HoursRev {
///         HoursRev(self)
///     }
///
///     /// Since `Clone::clone` isn't const callable on stable,
///     /// clonable iterators must define an inherent method to be cloned
///     const fn copy(&self) -> Self {
///         let Self{start, end} = *self;
///         Self{start, end}
///     }
/// }
///
/// struct HoursRev(Hours);
///
/// impl iter::ConstIntoIter for HoursRev {
///     type Kind = iter::IsIteratorKind;
///     type IntoIter = Self;
///     type Item = u8;
/// }
///
/// impl HoursRev {
///     const fn next(self) -> Option<(u8, Self)> {
///         konst::option::map!(self.0.next_back(), |(a, h)| (a, HoursRev(h)))
///     }
///
///     const fn next_back(self) -> Option<(u8, Self)> {
///         konst::option::map!(self.0.next(), |(a, h)| (a, HoursRev(h)))
///     }
///
///     const fn rev(self) -> Hours {
///         self.0
///     }
///
///     const fn copy(&self) -> Self {
///         Self(self.0.copy())
///     }
/// }
///
///
/// ```
///
#[doc(inline)]
pub use konst_kernel::into_iter::ConstIntoIter;

#[doc(no_inline)]
pub use crate::polymorphism::kinds::{IsIntoIterKind, IsIteratorKind, IsStdKind};

/// Trait for all the types that can be iterated over with ranges.
///
/// This trait is sealed and can only be implemented by `konst`
pub use konst_kernel::step_kk::Step;

include! {"./iter/collect_const.rs"}
include! {"./iter/iter_eval.rs"}
