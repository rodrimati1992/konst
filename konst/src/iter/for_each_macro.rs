#[doc(hidden)]
#[macro_export]
macro_rules! __for_each_hidden {
    (
        $pattern:pat in $iterator:expr
            $($(,)+ $iterator_method:ident ($($method_args:tt)*) )*
            $(,)?
        =>
        $($code:tt)*
    ) => ({
        $crate::iter::eval!{
            $iterator
                $(, $iterator_method ($($method_args)*) )*
                ,for_each(|$pattern| { $($code)* })
        }
    });
}

/// Iterates over all elements of an [iterator](crate::iter::ConstIntoIter),
/// const equivalent of [`Iterator::for_each`]
///
/// This macro supports emulating iterator methods by expanding to equivalent code.
/// They are documented in the [`iterator_dsl`] module,
/// because they are also supported by other `konst::iter` macros.
///
/// # Examples
///
/// ### Custom iterator
///
/// ```rust
/// use konst::iter::{ConstIntoIter, IsIteratorKind};
///
/// struct Upto10(u8);
///
/// impl ConstIntoIter for Upto10 {
///     type Kind = IsIteratorKind;
///     type IntoIter = Self;
///     type Item = u8;
///     const ITEMS_NEED_DROP: bool = false;
/// }
///
/// impl Upto10 {
///     const fn next(&mut self) -> Option<u8> {
///         if self.0 < 10 {
///             let ret = self.0;
///             self.0 += 1;
///             Some(ret)
///         } else {
///             None
///         }
///     }
/// }
///
/// const N: u32 = {
///     let mut n = 0u32;
///     konst::iter::for_each!{elem in Upto10(7) =>
///         n = n * 10 + elem as u32;
///     }
///     n
/// };
///
/// assert_eq!(N, 789);
///
/// ```
///
/// ### Summing pairs
///
// TODO: remove the ignore once array has by-value iteration
/// ```rust, ignore
/// use konst::iter::for_each;
///     
/// const fn add_pairs<const N: usize>(l: [u32; N], r: [u32; N]) -> [u32; N] {
///     let mut out = [0u32; N];
///
///     for_each!{(i, val) in l,zip(r),map(|(l, r)| l + r),enumerate() =>
///         out[i] = val;
///     }
///
///     out
/// }
///
/// assert_eq!(add_pairs([], []), []);
/// assert_eq!(add_pairs([3], [5]), [8]);
/// assert_eq!(add_pairs([3, 5], [8, 13]), [11, 18]);
///
/// ```
///
/// [`iterator_dsl`]: crate::iter::iterator_dsl
#[doc(inline)]
pub use __for_each_hidden as for_each;
