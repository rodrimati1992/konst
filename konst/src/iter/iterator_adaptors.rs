/// Const analog of [`core::iter::repeat`],
/// except that this requires the repeated value to impl `Copy`
/// (instead of `Clone`).
///
/// # Example
///
/// ```rust
/// use konst::iter::{self, collect_const};
///
/// const ARR: &[u8] = &collect_const!(u8 => iter::repeat(3),take(5));
///
/// assert_eq!(ARR, &[3, 3, 3, 3, 3]);
/// ```
pub use konst_kernel::iter::iter_adaptors::repeat;

/// Const analog of [`core::iter::Repeat`],
/// constructed by [`repeat`](crate::iter::repeat).
pub use konst_kernel::iter::iter_adaptors::Repeat;
