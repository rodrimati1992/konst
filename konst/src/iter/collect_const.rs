use typewit::{MakeTypeWitness, TypeEq, TypeWitnessTypeArg};

#[doc(hidden)]
#[macro_export]
macro_rules! __collect_const_hidden {
    (
        $Item:ty =>
        $($rem:tt)*
    ) => {{
        $crate::__collect_const_iter_with!{
            $Item ,
            {let item = $crate::__::MaybeUninit::new(item);},
            |array, length, item| array[length] = item,
            elem_length = 1,
            =>
            $($rem)*
        }

        __ARR81608BFNA5
    }}
}

/**
Collects an iterator constant into an array

# Drop behavior

The behavior regarding dropping iterators is
[documented here](crate::iter::ConstIntoIter#dropping).

# Iterator methods

This macro supports emulating iterator methods by expanding to equivalent code.

The supported iterator methods are documented in the [`iterator_dsl`] module,
because they are also supported by other `konst::iter` macros.

# Syntax

The syntax of this macro is:

```text
collect_const!(
    $Item:ty => $into_iterator:expr
        $(, $iterator_method:ident ($($method_args:tt)*) )*
        $(,)?
)
```
Where `$Item` is the type of the elements that'll be collected into an array.

Where `$into_iterator` is any type that can be converted into a const iterator, with
[`konst::iter::into_iter`](crate::iter::into_iter).

Where `$iterator_method` is any of the supported methods described in
the [`iterator_dsl`] module.

# Examples

### Iterating over a range

```rust
use konst::iter;

const ARR: [u64; 8] = iter::collect_const!(u64 =>
    10..,
        filter(|n| *n % 2 == 0),
        skip(5),
        take(8),
);

assert_eq!(ARR, [20, 22, 24, 26, 28, 30, 32, 34]);
```

### Iterating over an array

```rust
use konst::iter;

const ARR: [u8; 6] = iter::collect_const!(u8 =>
    [10, 20, 30],flat_map(|n| [n - 1, n + 1]),
);

assert_eq!(ARR, [9, 11, 19, 21, 29, 31]);
```



[`iterator_dsl`]: crate::iter::iterator_dsl
*/
#[doc(inline)]
pub use __collect_const_hidden as collect_const;

#[doc(hidden)]
#[macro_export]
macro_rules! __collect_const_iter_with {
    (
        $Item:ty,
        {$($reassign_item:tt)*},
        |$array:ident, $length:ident, $item:ident| $elem_initer:expr,
        elem_length = $elem_length:expr,
        =>
        $($rem:tt)*
    ) => {
        const fn __func_zxe7hgbnjs<Ret_KO9Y329U2U, const CAP_KO9Y329U2U: $crate::__::usize>(
            cmd: $crate::__::CollectorCmd<Ret_KO9Y329U2U, $Item, CAP_KO9Y329U2U>
        ) -> Ret_KO9Y329U2U {
            let mut $array = $crate::maybe_uninit::uninit_array::<_, CAP_KO9Y329U2U>();
            let mut $length = 0usize;

            $crate::iter::eval!{
                $($rem)*
                ,for_each(|$item| {
                    $($reassign_item)*
                    if let $crate::__::CollectorCmd::BuildArray(teq) = cmd {
                        teq.reachability_hint(());

                        $elem_initer
                    }

                    $length += $elem_length;
                })
            }

            match cmd {
                $crate::__::CollectorCmd::ComputeLength(teq) => {
                    $crate::__::forget($array);
                    teq.to_right($length)
                }
                $crate::__::CollectorCmd::BuildArray(teq) => {
                    $crate::__::assert!{
                        $length == CAP_KO9Y329U2U,
                        "initialization was skipped somehow",
                    }

                    // SAFETY: The above assert ensures that
                    // all of the array is initialized
                    let $array = unsafe{ $crate::maybe_uninit::array_assume_init($array) };
                    teq.to_right($array)
                }
            }
        }

        const __COUNT81608BFNA5: $crate::__::usize =
            __func_zxe7hgbnjs($crate::__::MakeTypeWitness::MAKE);

        const __ARR81608BFNA5: [$Item; __COUNT81608BFNA5] =
            __func_zxe7hgbnjs($crate::__::MakeTypeWitness::MAKE);


    };
}

#[doc(hidden)]
pub enum CollectorCmd<Ret, T, const CAP: usize> {
    ComputeLength(TypeEq<usize, Ret>),
    BuildArray(TypeEq<[T; CAP], Ret>),
}

impl<Ret, T, const CAP: usize> TypeWitnessTypeArg for CollectorCmd<Ret, T, CAP> {
    type Arg = Ret;
}

impl<T> MakeTypeWitness for CollectorCmd<usize, T, 0> {
    const MAKE: Self = Self::ComputeLength(TypeEq::NEW);
}

impl<T, const CAP: usize> MakeTypeWitness for CollectorCmd<[T; CAP], T, CAP> {
    const MAKE: Self = Self::BuildArray(TypeEq::NEW);
}
