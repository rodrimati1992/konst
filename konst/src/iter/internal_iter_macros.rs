use crate::iter::ConstIntoIter;

use core::marker::PhantomData;

#[doc(hidden)]
#[macro_export]
macro_rules! iterator_shared {
    (
        is_forward = $is_forward:ident,
        $(is_copy = $is_copy:ident,)?
        $(is_drop = $is_drop:ident,)?
        item = $Item:ty,
        iter_forward = $Self:ty,
        $(iter_reversed = $Rev:path,)?
        next($self:ident) $next_block:block,
        $(next_back $next_back_block:block,)?
        fields = $fields:tt,
    ) => {
        $crate::__iterator_shared! {
            is_forward = $is_forward,
            is_copy = ($($is_copy)? true),
            is_drop = ($($is_drop)? false),
            item = $Item,
            iter_forward = $Self,
            $(iter_reversed = $Rev,)?
            next($self) $next_block,
            $(next_back $next_back_block,)?
            fields = $fields,
        }
    }
}

#[doc(hidden)]
#[macro_export]
macro_rules! __iterator_shared {
    (
        is_forward = $is_forward:ident,
        is_copy = ($is_copy:ident $($_is_copy:tt)*),
        is_drop = ($is_drop:ident $($_is_drop:tt)*),
        item = $Item:ty,
        iter_forward = $Self:ty,
        $(iter_reversed = $Rev:path,)?
        next($self:ident) $next_block:block,
        $(next_back $next_back_block:block,)?
        fields = $fields:tt,
    ) => {
        $crate::__choose_alt! {($is_copy) {
            /// Creates a clone of this iterator
            pub const fn copy(&self) -> Self {
                $crate::__iterator_shared_copy!{self $fields}
            }
        }}

        $(
            $crate::__iterator_shared_rev!{
                rev = $crate::__choose!($is_forward $Rev $Self),
                is_drop = $is_drop,
                $fields
            }
        )?

        /// Advances the iterator and returns the next value.
        #[track_caller]
        pub const fn next(&mut $self) -> Option<$Item> {
            $crate::__choose!{
                $is_forward
                $next_block
                $($next_back_block)?
            }
        }

        $(
            /// Removes and returns an element from the end of the iterator.
            #[track_caller]
            pub const fn next_back(&mut $self) -> Option<$Item> {
                $crate::__choose!{
                    $is_forward
                    $next_back_block
                    $next_block
                }
            }
        )?
    };
}

#[doc(hidden)]
#[macro_export]
macro_rules! __iterator_shared_copy {
    ( $self:ident { $( $field:ident $(.$copy_method:ident())? ),* $(,)?} ) => ({
        Self {
            $($field: $self.$field $(.$copy_method())?,)*
        }
    })
}

#[doc(hidden)]
#[macro_export]
macro_rules! __iterator_shared_rev {
    (
        rev = $Rev:ty,
        is_drop = $is_drop:ident,
        { $( $field:ident $(.$copy_method:ident())? ),* $(,)?}
    ) => (
        /// Reverses the iterator
        pub const fn rev(self) -> $Rev {
            $crate::__choose_alt! {($is_drop) {
                $crate::destructure!{Self {$($field),*} = self}
            } else {
                let Self {$($field),*} = self;
            }}

            type Type<T> = T;
            Type::<$Rev> {$($field),*}
        }
    )
}

#[doc(hidden)]
#[inline(always)]
pub const fn __get_item_ty<Iter>(_: &Iter) -> PhantomData<(Iter, Iter::Item)>
where
    Iter: ConstIntoIter,
{
    PhantomData
}

#[doc(hidden)]
#[inline(always)]
pub const fn __assert_item_ty<Iter>(_: &Iter::Item, _: PhantomData<(Iter, Iter::Item)>)
where
    Iter: ConstIntoIter,
{
}

#[doc(hidden)]
#[inline(always)]
pub const fn __items_needs_drop<Iter>(_: &Iter) -> bool
where
    Iter: ConstIntoIter,
{
    Iter::ITEMS_NEED_DROP
}

#[doc(hidden)]
pub const fn __infer_option_ty<T>(_: &Option<T>, _: &T) {}
