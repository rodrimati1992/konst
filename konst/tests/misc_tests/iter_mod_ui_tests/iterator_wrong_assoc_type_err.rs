use konst::iter::{ConstIntoIter, IsIteratorKind};

struct WrongItem;

impl ConstIntoIter for WrongItem {
    type Kind = IsIteratorKind;
    type IntoIter = Self;
    type Item = u8;
}

impl WrongItem {
    const fn next(self) -> Option<(u64, Self)> {
        None
    }
}

const _: () = {
    konst::iter::for_each!{_ in WrongItem => }
};


struct WrongIntoIter;

impl ConstIntoIter for WrongIntoIter {
    type Kind = IsIteratorKind;
    type IntoIter = konst::slice::IterCopied<'static, u8>;
    type Item = u8;
}

impl WrongIntoIter {
    const fn next(self) -> Option<(u8, Self)> {
        None
    }
}

const _: () = {
    konst::iter::for_each!{_ in WrongIntoIter => }
};



fn main() {}