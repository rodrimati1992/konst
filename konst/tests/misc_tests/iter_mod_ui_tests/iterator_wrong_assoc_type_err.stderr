error[E0271]: type mismatch resolving `<WrongItem as ConstIntoIter>::Item == u64`
  --> tests/misc_tests/iter_mod_ui_tests/iterator_wrong_assoc_type_err.rs:19:5
   |
19 |     konst::iter::for_each!{_ in WrongItem => }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<WrongItem as ConstIntoIter>::Item == u64`
   |
note: expected this to be `u64`
  --> tests/misc_tests/iter_mod_ui_tests/iterator_wrong_assoc_type_err.rs:8:17
   |
8  |     type Item = u8;
   |                 ^^
   = note: this error originates in the macro `$crate::__iter2_interpreter` which comes from the expansion of the macro `konst::iter::for_each` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0271]: type mismatch resolving `<WrongIntoIter as ConstIntoIter>::IntoIter == WrongIntoIter`
  --> tests/misc_tests/iter_mod_ui_tests/iterator_wrong_assoc_type_err.rs:39:5
   |
39 |     konst::iter::for_each!{_ in WrongIntoIter => }
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type mismatch resolving `<WrongIntoIter as ConstIntoIter>::IntoIter == WrongIntoIter`
   |
note: expected this to be `WrongIntoIter`
  --> tests/misc_tests/iter_mod_ui_tests/iterator_wrong_assoc_type_err.rs:27:21
   |
27 |     type IntoIter = konst::slice::IterCopied<'static, u8>;
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: expected struct `WrongIntoIter`
              found struct `IterCopied<'static, u8>`
note: required by a bound in `IntoIterWrapper::<T, IsIteratorKind>::const_into_iter`
  --> src/iter.rs
   |
   |     pub const fn const_into_iter(self) -> T
   |                  --------------- required by a bound in this associated function
   |     where
   |         T: ConstIntoIter<IntoIter = T>,
   |                          ^^^^^^^^^^^^ required by this bound in `IntoIterWrapper::<T, IsIteratorKind>::const_into_iter`
   = note: this error originates in the macro `$crate::iter::into_iter` which comes from the expansion of the macro `konst::iter::for_each` (in Nightly builds, run with -Z macro-backtrace for more info)
